### PR TITLE
Better: Add InvalidConfiguration error code to allow custom message or behavior.

### PR DESCRIPTION
**Pull request by aymebou**
Hey !

- Ticket(s): RD-18471 (that's for @CedricBm)

In the effort of translating our application, we would like to be able to translate the error messages raised when configuration is invalid.

What I propose is to have a `InvalidConfiguration.error_code` indicating what is the error along with `InvalidConfiguration.data` containing the data related to the error.

I hesitated storing the whole record inside the error, it would make accessing the data simpler but it seemed like a waste of storage - might be irrelevant since it's a pointer. Anyway let me know if you disagree.

By the way, I can confirm all the messages have a corresponding spec so I didn't have to add any.

### CODE REVIEW

**Comment by aymebou** [comment]
For the record, for our Application the goal is to use the error_code this way :

rescue WorkingHours::InvalidConfiguration => e
    errors.add(:base, I18n.t(e.error_code, scope: "working_hours.invalid_configuration")
  end

**Review by Intrepidd — APPROVED** [review]
Great stuff, love the backward compatibility

File: lib/working_hours/config.rb [hunk 1]
```diff
@@ -1,7 +1,35 @@
 require 'set'
 
 module WorkingHours
-  InvalidConfiguration = Class.new StandardError
+  class InvalidConfiguration < StandardError
+    attr_reader :data, :error_code
+
+    def initialize(error_code, data: {})
+      @data = data unless data.blank?
```
Comment (jarthod) [inline lib/working_hours/config.rb:8]: Quick comment here: not sure why we default data to `{}` if we don't store it when blank? why not default data to `nil` in this case (and remove the `unless`)?

**Review by jarthod — APPROVED** [review]
Looks pretty good to me, thanks for the Readme update.

I tend to prefer the way you passed only the needed objects as `data`, it's safer to avoid memory leaks and stuff like that. Passing the whole config object would not necessarily consume memory, but if the exception ends-up not being garbage collected because it's references somewhere we'll never be able to GC the config object.

File: lib/working_hours/config.rb (hunk 1 above)
Comment (aymebou) [reply lib/working_hours/config.rb:8]: Oh yeah good idea, it's indeed simpler

**Comment by jarthod** [comment]
Thanks! @aymebou @CedricBm any other MR to expect soon or shall we cut a release with only these changes? Where you able to translate all the errors needed with this?

**Comment by aymebou** [comment]
> Thanks! @aymebou @CedricBm any other MR to expect soon or shall we cut a release with only these changes? Where you able to translate all the errors needed with this?

Actually could you wait a bit ? PR on the other side's not fully reviewed yet, just want to make sure I'm not forgetting anything

**Comment by jarthod** [comment]
Yep sure, let us know once it's complete

File: README.md [hunk 2]
```diff
@@ -116,6 +116,21 @@ If *any* hours are set for a calendar day in `holiday_hours`, then the `working_
...
+
+You can also access the error code in case you want to implement custom behavior or changing one specific message, e.g:
+
+```ruby
+rescue WorkingHours::InvalidConfiguration => e
+    if e.error_code == :empty
+      raise StandardError.new "Config is required"
+    end
+    raise e
+end
```
Comment (CedricBm) [inline README.md:131]: There's an indentation issue here. First one is 4 spaces, second is 2.

**Review by CedricBm — COMMENTED** [review]
It looks good indeed. We'll ping you once we're ready on the other side :smile:

**Comment by aymebou** [comment]
Code is approved on our side, you can merge this PR and release when you have the time.

Thank you for your patience !

**Comment by jarthod** [comment]
Merged and released in `1.4.1` :tada:

**Pull request merged**